network/p2p/scoring/registry.go

```diff
@@ -401,12 +417,24 @@ func (r *GossipSubAppSpecificScoreRegistry) subscriptionPenalty(pid peer.ID, flo
 	return 0
 }
 
+// duplicateMessagesPenalty returns the duplicate message penalty for a peer.
+func (r *GossipSubAppSpecificScoreRegistry) duplicateMessagesPenalty(pid peer.ID) float64 {
+	duplicateMessageCount := r.getDuplicateMessageCount(pid)
+	if duplicateMessageCount != 0 {
```

Comment:
In addition to `DefaultDuplicateMessagePenalty = -10e-4`, we should also define a `DefaultDuplicateMessageThreshold = 10e+4`. Based on the current approach on tracking the duplicate message delivery using the GossipSub tracer, the counter value does not necessarily indicate a misbehavior itself. The GossipSub will count every message that is _not fresh_ as a duplicate. So all gossips for the same message except the first one are considered duplicates. 

We should treat duplicate message count of a node as a misbehavior only when it stands as an anomaly. To decide whether the count is anomaly or not, we define the threshold, and we set to a high value (10+e4). Accordingly, here, we should check whether the `duplicateMessageCount > threshold`, and if it is, we should penalize the node for the _excess_ amount floored at the `MaxAppSpecificPenalty` (as it is now). 

With a regular decay of 0.5, the small penalty of `-10e-4` and a large threshold of `10e4`, the normal behavior of the nodes are set safe, while anomalies are treated as misbehavior and penalized with the excess.

Please also add test ensuring that as long as the node stays below the threshold on duplicate message delivery, it is not penalized.

Reply:
https://github.com/onflow/flow-go/pull/5058/commits/59462cd9a69895cfdb26af133d3d8d7aafe64db3